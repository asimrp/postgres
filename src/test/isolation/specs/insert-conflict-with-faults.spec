# INSERT ... ON CONFLICT test verifying that speculative insertion
# failures are handled

setup
{
    CREATE TABLE upserttest(key text, data text);
    CREATE UNIQUE INDEX ON upserttest(key);

    CREATE EXTENSION faultinjector;
    -- start with a clean slate
    SELECT * FROM inject_fault('all', 'reset');

    -- inject fault to suspend insert transaction after a tuple has
    -- been inserted into the heap but before it is inserted into the
    -- index.
    SELECT * FROM inject_fault('insert_index_tuples', 'suspend');
    SELECT * FROM inject_fault('heap_abort_speculative', 'skip');
}

teardown
{
    DROP TABLE upserttest;
    SELECT * FROM inject_fault('all', 'reset');
}


session "s1"
setup
{
  SET default_transaction_isolation = 'read committed';
}
step "s1_upsert" { INSERT INTO upserttest(key, data) VALUES('k1', 'inserted s1') ON CONFLICT (key) DO UPDATE SET data = upserttest.data || ' with conflict update s1'; }
step "s1_fault_status" { SELECT * FROM inject_fault('heap_abort_speculative', 'status'); }
step "s1_select" { SELECT * FROM upserttest; }

session "s2"
setup
{
  SET default_transaction_isolation = 'read committed';
}
step "s2_upsert" { INSERT INTO upserttest(key, data) VALUES('k1', 'inserted s2') ON CONFLICT (key) DO UPDATE SET data = upserttest.data || ' with conflict update s2'; }
step "s2_unblock_s1" { SELECT * FROM inject_fault('insert_index_tuples', 'resume'); }

# Test that speculative locks are correctly acquired and released, s2
# inserts, s1 updates.
permutation
   # S1 should hit the fault and block
   "s1_upsert"&
   # S2 should insert without conflict
   "s2_upsert"
   "s2_unblock_s1"
   "s1_fault_status"
   "s1_select"
