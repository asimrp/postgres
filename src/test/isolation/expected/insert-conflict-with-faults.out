Parsed test spec with 2 sessions

starting permutation: s1_upsert s2_upsert s2_unblock_s1 s1_fault_status s1_select
inject_fault   

Success:       
step s1_upsert: INSERT INTO upserttest(key, data) VALUES('k1', 'inserted s1') ON CONFLICT (key) DO UPDATE SET data = upserttest.data || ' with conflict update s1'; <waiting ...>
step s2_upsert: INSERT INTO upserttest(key, data) VALUES('k1', 'inserted s2') ON CONFLICT (key) DO UPDATE SET data = upserttest.data || ' with conflict update s2';
step s2_unblock_s1: SELECT * FROM inject_fault('insert_index_tuples', 'resume');
inject_fault   

Success:       
step s1_upsert: <... completed>
step s1_fault_status: SELECT * FROM inject_fault('heap_abort_speculative', 'status');
inject_fault   

Success: fault name:'heap_abort_speculative' fault type:'skip' database name:'' table name:'' start occurrence:'1' end occurrence:'1' extra arg:'0' fault injection state:'completed' num times hit:'1' 

step s1_select: SELECT * FROM upserttest;
key            data           

k1             inserted s2 with conflict update s1
inject_fault   

Success:       
